Mathlib/Probability/Kernel/Disintegration/Density.lean

```diff
@@ -0,0 +1,799 @@
+    change Measurable ((fun (p : α × _) ↦ κ p.1 (countablePartitionSet n p.2 ×ˢ s)
+      / ν p.1 (countablePartitionSet n p.2)) ∘ (fun x ↦ (a, x)))
+    exact (measurable_densityProcess_aux κ ν n hs).comp measurable_prod_mk_left
+  · refine ae_of_all _ (fun x ↦ ?_)
+    by_cases h0 : ν a (countablePartitionSet n x) = 0
+    · suffices κ a (countablePartitionSet n x ×ˢ s) = 0 by simp [h0, this]
+      have h0' : fst κ a (countablePartitionSet n x) = 0 :=
+        le_antisymm ((hκν a _).trans h0.le) zero_le'
+      rw [fst_apply' _ _ (measurableSet_countablePartitionSet _ _)] at h0'
+      refine measure_mono_null (fun x ↦ ?_) h0'
+      simp only [mem_prod, mem_setOf_eq, and_imp]
+      exact fun h _ ↦ h
+    · exact ENNReal.div_lt_top (measure_ne_top _ _) h0
+  congr
+  have : ∫⁻ x in u, κ a (countablePartitionSet n x ×ˢ s) / ν a (countablePartitionSet n x) ∂(ν a)
+      = ∫⁻ _ in u, κ a (u ×ˢ s) / ν a u ∂(ν a) := by
+    refine set_lintegral_congr_fun hu_meas (ae_of_all _ (fun t ht ↦ ?_))
+    rw [countablePartitionSet_of_mem hu ht]
+  rw [this]
+  simp only [MeasureTheory.lintegral_const, MeasurableSet.univ, Measure.restrict_apply, univ_inter]
+  by_cases h0 : ν a u = 0
+  · simp only [h0, mul_zero]
+    have h0' : fst κ a u = 0 := le_antisymm ((hκν a _).trans h0.le) zero_le'
+    rw [fst_apply' _ _ hu_meas] at h0'
+    refine (measure_mono_null ?_ h0').symm
+    intro p
+    simp only [mem_prod, mem_setOf_eq, and_imp]
+    exact fun h _ ↦ h
+  rw [div_eq_mul_inv, mul_assoc, ENNReal.inv_mul_cancel h0, mul_one]
+  exact measure_ne_top _ _
+
+lemma integral_densityProcess (hκν : fst κ ≤ ν) [IsFiniteKernel κ] [IsFiniteKernel ν]
+    (n : ℕ) (a : α) {s : Set β} (hs : MeasurableSet s) :
+    ∫ x, densityProcess κ ν n a x s ∂(ν a) = (κ a (univ ×ˢ s)).toReal := by
+  rw [← integral_univ, ← sUnion_countablePartition γ n, sUnion_eq_iUnion,
+    iUnion_prod_const, measure_iUnion]
+  rotate_left
+  · intro i j hij
+    simp only [Set.disjoint_prod, disjoint_self, bot_eq_empty]
+    refine Or.inl (disjoint_countablePartition i.prop j.prop ?_)
+    rw [ne_eq, Subtype.coe_inj]
+    exact hij
+  · exact fun k ↦ (measurableSet_countablePartition n k.prop).prod hs
+  rw [integral_iUnion]
+  rotate_left
+  · exact fun k ↦ measurableSet_countablePartition n k.prop
+  · refine fun i j hij ↦ disjoint_countablePartition i.prop j.prop ?_
+    rw [ne_eq, Subtype.coe_inj]
+    exact hij
+  · exact (integrable_densityProcess hκν n a hs).integrableOn
+  rw [ENNReal.tsum_toReal_eq (fun _ ↦ measure_ne_top _ _)]
+  congr with k
+  rw [set_integral_densityProcess_I hκν _ _ hs k.prop]
+
+lemma set_integral_densityProcess (hκν : fst κ ≤ ν) [IsFiniteKernel κ] [IsFiniteKernel ν]
+    (n : ℕ) (a : α) {s : Set β} (hs : MeasurableSet s) {A : Set γ}
+    (hA : MeasurableSet[countableFiltration γ n] A) :
+    ∫ x in A, densityProcess κ ν n a x s ∂(ν a) = (κ a (A ×ˢ s)).toReal := by
+  refine induction_on_inter (m := countableFiltration γ n) (s := countablePartition γ n)
```

Comment:
It seems a little bit far-fetched to use a pi-system argument in such a simple situation. Do you have the fact that any set measurable wrt `countableFiltration γ n` is a finite union of atoms? From this, the conclusion would be direct from `set_integral_densityProcess_I`, without the need for ` integral_densityProcess`.